In [1]:
%matplotlib inline
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from IPython.display import Image
sns.set()

In [2]:
pip install matplotlib_venn --upgrade --user

https://www.glerl.noaa.gov/metdata/mil/archive/ 

In [3]:
pip install pandas --upgrade --user

Note: you may need to restart the kernel to use updated packages.


In [4]:
pip install numexpr --upgrade --user

Note: you may need to restart the kernel to use updated packages.


In [5]:
pip install seaborn --upgrade --user

In [6]:
pip install scipy --upgrade --user

  Attempting uninstall: scipy
    Found existing installation: scipy 1.7.3
    Uninstalling scipy-1.7.3:
      Successfully uninstalled scipy-1.7.3
Note: you may need to restart the kernel to use updated packages.


ERROR: Could not install packages due to an OSError: [WinError 5] Access is denied: 'C:\\Users\\fassg\\AppData\\Roaming\\Python\\Python39\\site-packages\\~cipy\\.libs\\libansari.54HGNEJBQIYZX5TZPCQGLNVIPFU6NWEX.gfortran-win_amd64.dll'
Check the permissions.



In [7]:
original_data = pd.read_csv("mkecallswheader.csv")

In [8]:
original_data.head(10)

id                 time                     location district  \
0  2093116  2019-05-21 15:19:03      7420 W GOOD HOPE RD,MKE        4   
1  2093127  2019-05-21 15:24:30           1421 N 27TH ST,MKE        3   
2  2093141  2019-05-21 15:25:46           4054 N 71ST ST,MKE        7   
3  2093805  2019-05-21 20:46:28         245 W LINCOLN AV,MKE        2   
4  2093816  2019-05-21 20:50:03          1721 W CANAL ST,MKE        3   
5  2093829  2019-05-21 21:02:37  E WRIGHT ST / N WEIL ST,MKE        5   
6  2093872  2019-05-21 20:50:47           9010 N 95TH ST,MKE        4   
7  2093887  2019-05-21 21:25:33          983 W ARTHUR AV,MKE        2   
8  2093918  2019-05-21 21:36:05           4115 N 56TH ST,MKE        7   
9  2093929  2019-05-21 21:45:53        7806 W HAMPTON AV,MKE        7   

            nature                        status          inserted_at  \
0           ACC PI           Service in Progress  2019-05-21 20:51:09   
1     TRAFFIC STOP       City Citation(s) Issued  2019-05-21 20:57:11   
2      SUBJ WANTED          Assignment Completed  2019-05-21 21:00:12   
3   SPECIAL ASSIGN           Service in Progress  2019-05-22 02:22:32   
4      TRBL W/SUBJ  Unable to Locate Complainant  2019-05-22 02:25:33   
5    PARK AND WALK           Service in Progress  2019-05-22 02:37:36   
6  WELFARE CITIZEN                       Advised  2019-05-22 02:52:43   
7       BATTERY DV           Service in Progress  2019-05-22 03:01:48   
8   RETURN STATION          Assignment Completed  2019-05-22 03:16:51   
9     TRAFFIC STOP                       Advised  2019-05-22 03:22:52   

            updated_at                                              point  \
0  2019-05-21 20:51:09  0101000020E6100000FC7C94111793454061D971683600...   
1  2019-05-21 20:57:11  0101000020E6100000D2AB014A4386454067C416CCA9FC...   
2  2019-05-21 21:00:12  0101000020E610000053FFC5D8AE8B45402CAE3B270700...   
3  2019-05-22 02:22:32  0101000020E610000078ABF8D04F804540633ABE0779FA...   
4  2019-05-22 02:25:33  0101000020E6100000E8323509DE834540C3D7D7BAD4FB...   
5  2019-05-22 02:37:36  0101000020E6100000DEF1DC312B88454059D878558CF9...   
6  2019-05-22 02:52:43  0101000020E6100000357D76C07597454080B4FF01D601...   
7  2019-05-22 03:01:48  0101000020E6100000BC033C69E17F454041ABDDC02EFB...   
8  2019-05-22 03:16:51  0101000020E610000014483FD0C08B45404EB747CAF1FE...   
9  2019-05-22 03:22:52  0101000020E6100000F6F0C05B7B8D45404F34B4A69E00...   

     call_id  
0  191411633  
1  191411672  
2  191411674  
3  191412545  
4  191412465  
5  191412584  
6  191412544  
7  191412632  
8  191412656  
9  191412676

In [98]:
original_data.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4027695 entries, 0 to 4027694
Data columns (total 10 columns):
 #   Column       Non-Null Count    Dtype 
---  ------       --------------    ----- 
 0   id           4027695 non-null  int64 
 1   time         4027695 non-null  object
 2   location     4027695 non-null  object
 3   district     3937463 non-null  object
 4   nature       4027695 non-null  object
 5   status       4027695 non-null  object
 6   inserted_at  4027695 non-null  object
 7   updated_at   4027695 non-null  object
 8   point        3958766 non-null  object
 9   call_id      4027695 non-null  int64 
dtypes: int64(2), object(8)
memory usage: 307.3+ MB


In [10]:
original_data.describe()

id       call_id
count  4.027695e+06  4.027695e+06
mean   5.589211e+06  1.891808e+08
std    2.392488e+07  1.459309e+07
min    1.000000e+00  1.630815e+08
25%    1.006924e+06  1.732129e+08
50%    2.013848e+06  1.906707e+08
75%    3.020772e+06  2.016115e+08
max    1.635417e+08  2.201010e+08

From the above calls to the [`.head()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.head.html), [`.info()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html), and [`.describe()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html) methods it is possible to see that there are $10$ total features. Of these features there are two features that are formatted as integers and eight features formatted as the default object that pandas imports non-numerical features as. From the object classified features there are a few that can be converted to new types. The time, inserted_at, and updated_at column should be formatted as a date time object. District should be converted to a numerical categorical value. Nature and status should be converted to categorical features. Location should be kept as an object feature. More features should be extracted from the location in order to draw further observations. Both street name and street suffix would be good features to extract. The ID and point features are artifacts of the database the police dataset came from and as such will be dropped.

The first step is to copy the data into a new dataframe so that the original data can be preserved

In [11]:
df = original_data.copy(deep=True)

Now convert the time, inserted_at, and updated_at features into date time objects.

In [12]:
df['time'] = pd.to_datetime(df['time'], infer_datetime_format=True)  
df['inserted_at'] = pd.to_datetime(df['inserted_at'], infer_datetime_format=True)  
df['updated_at'] = pd.to_datetime(df['updated_at'], infer_datetime_format=True)

In [13]:
df.head()

id                time                 location district  \
0  2093116 2019-05-21 15:19:03  7420 W GOOD HOPE RD,MKE        4   
1  2093127 2019-05-21 15:24:30       1421 N 27TH ST,MKE        3   
2  2093141 2019-05-21 15:25:46       4054 N 71ST ST,MKE        7   
3  2093805 2019-05-21 20:46:28     245 W LINCOLN AV,MKE        2   
4  2093816 2019-05-21 20:50:03      1721 W CANAL ST,MKE        3   

           nature                        status         inserted_at  \
0          ACC PI           Service in Progress 2019-05-21 20:51:09   
1    TRAFFIC STOP       City Citation(s) Issued 2019-05-21 20:57:11   
2     SUBJ WANTED          Assignment Completed 2019-05-21 21:00:12   
3  SPECIAL ASSIGN           Service in Progress 2019-05-22 02:22:32   
4     TRBL W/SUBJ  Unable to Locate Complainant 2019-05-22 02:25:33   

           updated_at                                              point  \
0 2019-05-21 20:51:09  0101000020E6100000FC7C94111793454061D971683600...   
1 2019-05-21 20:57:11  0101000020E6100000D2AB014A4386454067C416CCA9FC...   
2 2019-05-21 21:00:12  0101000020E610000053FFC5D8AE8B45402CAE3B270700...   
3 2019-05-22 02:22:32  0101000020E610000078ABF8D04F804540633ABE0779FA...   
4 2019-05-22 02:25:33  0101000020E6100000E8323509DE834540C3D7D7BAD4FB...   

     call_id  
0  191411633  
1  191411672  
2  191411674  
3  191412545  
4  191412465

In [99]:
# https://stackoverflow.com/a/43427677 for arguments to get null counts
df.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4027695 entries, 0 to 4027694
Data columns (total 9 columns):
 #   Column       Non-Null Count    Dtype         
---  ------       --------------    -----         
 0   time         4027695 non-null  datetime64[ns]
 1   location     4027695 non-null  object        
 2   district     3937463 non-null  category      
 3   nature       4027695 non-null  category      
 4   status       4027695 non-null  category      
 5   inserted_at  4027695 non-null  datetime64[ns]
 6   updated_at   4027695 non-null  datetime64[ns]
 7   point        3958766 non-null  object        
 8   call_id      4027695 non-null  int64         
dtypes: category(3), datetime64[ns](3), int64(1), object(2)
memory usage: 199.8+ MB


In [15]:
df.describe()

id       call_id
count  4.027695e+06  4.027695e+06
mean   5.589211e+06  1.891808e+08
std    2.392488e+07  1.459309e+07
min    1.000000e+00  1.630815e+08
25%    1.006924e+06  1.732129e+08
50%    2.013848e+06  1.906707e+08
75%    3.020772e+06  2.016115e+08
max    1.635417e+08  2.201010e+08

From the above calls to the [`.head()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.head.html), [`.info()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html), and [`.describe()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html) methods it is possible to see that the time base features are now properly formatted. The next step is to bin the features by hour. This will be done in order to reduce the granularity of the data as well as matching up this dataset with hourly weather data.

In [16]:
# TODO

Next the id and point columns will be dropped since they are artifacts of the database the data was stored in. The point column is a hexadecimal postgres geometry point object that matches up to a location. Since it can not be reverse interpreted and will be correlated with location we will be dropping it.

In [17]:
df = df.drop('id', axis=1)
# df = df.drop('point', axis=1) ### do not want to drop this as its a postgres geospatial for approx lat long.
df.head()

time                 location district          nature  \
0 2019-05-21 15:19:03  7420 W GOOD HOPE RD,MKE        4          ACC PI   
1 2019-05-21 15:24:30       1421 N 27TH ST,MKE        3    TRAFFIC STOP   
2 2019-05-21 15:25:46       4054 N 71ST ST,MKE        7     SUBJ WANTED   
3 2019-05-21 20:46:28     245 W LINCOLN AV,MKE        2  SPECIAL ASSIGN   
4 2019-05-21 20:50:03      1721 W CANAL ST,MKE        3     TRBL W/SUBJ   

                         status         inserted_at          updated_at  \
0           Service in Progress 2019-05-21 20:51:09 2019-05-21 20:51:09   
1       City Citation(s) Issued 2019-05-21 20:57:11 2019-05-21 20:57:11   
2          Assignment Completed 2019-05-21 21:00:12 2019-05-21 21:00:12   
3           Service in Progress 2019-05-22 02:22:32 2019-05-22 02:22:32   
4  Unable to Locate Complainant 2019-05-22 02:25:33 2019-05-22 02:25:33   

                                               point    call_id  
0  0101000020E6100000FC7C94111793454061D971683600...  191411633  
1  0101000020E6100000D2AB014A4386454067C416CCA9FC...  191411672  
2  0101000020E610000053FFC5D8AE8B45402CAE3B270700...  191411674  
3  0101000020E610000078ABF8D04F804540633ABE0779FA...  191412545  
4  0101000020E6100000E8323509DE834540C3D7D7BAD4FB...  191412465

The nature, status, and district features will now be converted into categorical features

In [18]:
df['district'].unique()

array(['4', '3', '7', '2', '5', '6', '1', nan, 'OCOE', 'OUT', 'NTF',
       'SPD', 'CITY', 'ICS3', 'CIB', 'DPR', 'TRU', 'SCD', 'ICS', 'SF',
       'NLA', 'ICS1', 'IFC', 'FI', 'SID', 'D0', 'ICS7', 'ICS5', 'JUNE',
       'DDAC', 'MIRT', 'ICS6', 'ID', 'ADMN', 'TEU', 'MID', 'NID'],
      dtype=object)

In [19]:
df['district'].value_counts()

3       673002
7       650109
5       567958
4       552701
2       545253
6       473525
1       448786
CITY      7080
NTF       6274
SPD       5477
OCOE      3316
OUT       2249
DDAC       399
SCD        300
DPR        277
ICS1       229
CIB        218
FI          83
SF          55
NLA         34
ICS3        32
SID         31
ICS         23
TEU         16
ICS5         9
JUNE         5
ICS6         4
IFC          4
TRU          3
D0           2
MIRT         2
ADMN         2
NID          2
ICS7         1
ID           1
MID          1
Name: district, dtype: int64

Running a value counts on the different districts shows that there are more than the expected number of police districts in the city of Milwaukee present in the data. There should only be districts one through seven. Instead the data contains more districts than expected. These districts will be converted into categorical anyways as their entries will be useful for some observations. The erraneous districts will likely be ignored when drawing district based conclusions as it is not known what the other districts mean. District will be converted to a categorical with 36 different categories.

In [20]:
df['nature'].unique()

array(['ACC PI', 'TRAFFIC STOP', 'SUBJ WANTED', 'SPECIAL ASSIGN',
       'TRBL W/SUBJ', 'PARK AND WALK', 'WELFARE CITIZEN', 'BATTERY DV',
       'RETURN STATION', 'BUSINESS CHECK', 'SUSPICIOUS-OTH', 'REPORTS',
       'THEFT', 'VIOL REST ORDER', 'SUBJ WITH GUN', 'TAVERN CHECK',
       'PATROL', 'FOLLOW UP', 'ACC PDO', 'TRAFFIC HAZARD',
       'FAMILY TROUBLE', 'NON PURSUIT', 'BATTERY', 'INVESTIGATION',
       'PRISONER TRANS', 'BUS INV', 'CALL FOR POLICE', 'THEFT VEHICLE',
       'SUSP PERS/AUTO', 'ENTRY TO AUTO', 'ASSIGNMENT', 'PROPERTY DAMAGE',
       'IND EXPOSURE', 'NOISE NUISANCE', 'PROPERTY PICKUP', 'SOLICITING',
       'STOLEN VEHICLE', 'OUT OF SERVICE', 'COURT DUTY', 'ENTRY',
       'PARKING TROUBLE', 'GRAFFITI', 'TRBL W/JUV', 'SHOTS FIRED',
       'SHOTSPOTTER', 'BUS INVESTIGATIO', 'THREAT', 'SUBJ WITH WEAPON',
       'TRAFFIC LASER', 'HOME VISIT DV', 'COMMUNITY MTNG',
       'CITIZEN CONTACT', 'RECK USE OF WEAP', 'ABAND/STOLEN PRO',
       'ASSIGN-ADMN MPD', 'CONVEY PROPERTY',

In [21]:
df['nature'].value_counts()

TRAFFIC STOP        401644
BUSINESS CHECK      325863
TRBL W/SUBJ         237324
RETURN STATION      209934
FOLLOW UP           150661
                     ...  
ACC PD1                  1
ABAND/LOST               1
2831 N 21ST              1
THREAT TO SCHOOL         1
STOLEN PROP              1
Name: nature, Length: 317, dtype: int64

The natures all look normal. They will be converted into a categorical value which will have 317 categories.

In [22]:
df['status'].unique()

array(['Service in Progress', 'City Citation(s) Issued',
       'Assignment Completed', 'Unable to Locate Complainant', 'Advised',
       'To be Filed', 'Advised/Referral', 'No Prosecution',
       'Open Investigation', 'Cleared by Arrest', 'False Alarm',
       'Filed Driver Exchange Report', 'Patrol Request', 'Referral',
       'Ordered to Appear', 'State Citation(s) Issued',
       'False Alarm (Weather Related)'], dtype=object)

In [23]:
df['status'].value_counts()

Service in Progress              1424568
Assignment Completed             1186663
Advised                           615111
Unable to Locate Complainant      336406
To be Filed                       164199
City Citation(s) Issued           153937
Advised/Referral                   87051
Open Investigation                 40803
No Prosecution                      6475
Cleared by Arrest                   4696
False Alarm                         2889
Filed Driver Exchange Report        2631
Referral                            1291
Patrol Request                       697
State Citation(s) Issued             160
False Alarm (Weather Related)         93
Ordered to Appear                     25
Name: status, dtype: int64

Status looks normal as well and will be converted to a categorical value. It will contain 17 different categories.

In [24]:
df['district'] = df['district'].astype("category")
df['nature'] = df['nature'].astype("category")
df['status'] = df['status'].astype('category')

In [25]:
df['district'].dtype

CategoricalDtype(categories=['1', '2', '3', '4', '5', '6', '7', 'ADMN', 'CIB', 'CITY',
                  'D0', 'DDAC', 'DPR', 'FI', 'ICS', 'ICS1', 'ICS3', 'ICS5',
                  'ICS6', 'ICS7', 'ID', 'IFC', 'JUNE', 'MID', 'MIRT', 'NID',
                  'NLA', 'NTF', 'OCOE', 'OUT', 'SCD', 'SF', 'SID', 'SPD',
                  'TEU', 'TRU'],
, ordered=False)

In [26]:
df['nature'].dtype

CategoricalDtype(categories=['.', '0', '1 BLOCK NORTH OF', '1301', '1359', '1603', '1733',
                  '230 N 37TH ST', '2532', '2831 N 21ST',
                  ...
                  'VIOL REST ORD', 'VIOL REST ORD-DV', 'VIOL REST ORDER',
                  'WATER MAIN BREAK', 'WATER MAIN BRK', 'WEAPON',
                  'WELFARE CHK', 'WELFARE CITIZEN', 'WIRES DOWN', '`'],
, ordered=False)

In [27]:
df['status'].dtype

CategoricalDtype(categories=['Advised', 'Advised/Referral', 'Assignment Completed',
                  'City Citation(s) Issued', 'Cleared by Arrest',
                  'False Alarm', 'False Alarm (Weather Related)',
                  'Filed Driver Exchange Report', 'No Prosecution',
                  'Open Investigation', 'Ordered to Appear', 'Patrol Request',
                  'Referral', 'Service in Progress',
                  'State Citation(s) Issued', 'To be Filed',
                  'Unable to Locate Complainant'],
, ordered=False)

The nature, status, and district features are now formatted as categories.

The last step is to extract features from the location category

In [54]:
def get_street_info(address: str) -> (int, str, str, int):
    """
    This method will take in a string representing an address and will return the information present in that address.
    Some example addresses are as follows:
        0             7420 W GOOD HOPE RD,MKE
        1                  1421 N 27TH ST,MKE
        2                  4054 N 71ST ST,MKE
        3                245 W LINCOLN AV,MKE
        4                 1721 W CANAL ST,MKE
        5         E WRIGHT ST / N WEIL ST,MKE
        6                  9010 N 95TH ST,MKE
    :param address: the string passed in representing the address
    :return: a tuple containing (home number, street name, street type)
    :auth: Grant Fass
    :since: 6 December 2021
    """
    street_type_lookup = ["ALY", "ANX", "ARC", "AVE", "BYU", "BCH", "BND", "BLF", "BLFS", "BTM", "BLVD", "BR", "BRG", "BRK", "BRKS", "BG", "BGS", "BYP", "CP", "CYN", "CPE",
                          "CSWY", "CTR", "CTRS", "CIR", "CIRS", "CLF", "CLFS", "CLB", "CMN", "CMNS", "COR", "CORS", "CRSE", "CT", "CTS", "CV", "CVS", "CRK", "CRES", "CRST", 
                          "XING", "XRD", "XRDS", "CURV", "DL", "DM", "DV", "DR", "EST", "ESTS", "EXPY", "EXT", "EXTS", "FALL", "FLS", "FRY", "FLD", "FLDS", "FLT", "FLTS", 
                          "FRD", "FRDS", "FRST", "FRG", "FRGS", "FRK", "FRKS", "FT", "FWY", "GDN", "GDNS", "GTWY", "GLN", "GLNS", "GRN", "GRNS", "GRV", "GRVS", "HBR", "HBRS", 
                          "HVN", "HTS", "HWY", "HL", "HLS", "HOLW", "INLT", "IS", "ISS", "ISLE", "JCT", "JCTS", "KY", "KYS", "KNL", "KNLS", "LK", "LKS", "LAND", "LNDG", "LN",
                          "LGT", "LGTS", "LF", "LCK", "LCKS", "LDG", "LOOP", "MALL", "MNR", "MNRS", "MDW", "MDWS", "MEWS", "ML", "MLS", "MSN", "MTWY", "MT", "MTN", "MTNS", 
                          "NCK", "ORCH", "OVAL", "OPAS", "PARK", "PKWY", "PASS", "PSGE", "PATH", "PIKE", "PNE", "PNES", "PL", "PLN", "PLNS", "PLZ", "PT", "PTS", "PRT", "PRTS", 
                          "PR", "RADL", "RAMP", "RNCH", "RPD", "RPDS", "RST", "RDG", "RDGS", "RIV", "RD", "RDS", "RTE", "ROW", "RUE", "RUN", "SHL", "SHLS", "SHR", "SHRS", 
                          "SKWY", "SPG", "SPGS", "SPUR", "SQ", "SQS", "STA", "STRA", "STRM", "ST", "STS", "SMT", "TER", "TRWY", "TRCE", "TRAK", "TRFY", "TRL", "TRLR", "TUNL",
                          "TPKE", "UPAS", "UN", "UNS", "VLY", "VLYS", "VIA", "VW", "VWS", "VLG", "VLGS", "VL", "VIS", "WALK", "WALL", "WAY", "WAYS", "WL", "WLS", "AV"]
    # this is used primarily for error checking
    unmatched_suffix = ""
    # remove the ,MKE suffix from the location if present
    address = address.removesuffix(",MKE")
    # Array containing the seperate addresses in the passed entry
    addresses = []
    # Check if the entry contains a / or not.
    # The presence of a / denotes the entry as a corner with two streets present
    # For example: N HUMBOLDT AV / E NORTH AV
    if ('/' in address):
        addresses = address.split(' / ')
    else:
        addresses = [address]
    
    # now perform opperations for each address
    # print(addresses)
    out = []
    for a in addresses:
        # Set up the values to be returned
        house_number = None
        street = None
        street_suffix = None
        # split appart the address on spaces
        s = a.split(' ')
        # Check if the first cell is a nueric. This would be the house number if it is a numeric
        if s[0].isnumeric():
            # house number present
            house_number = int(s[0])
            street = ' '.join(s[1:-1]) # use -1 since last index is exclusive
        else:
            street = ' '.join(s[0:-1])
        # update the street suffix based on the last entry in the array
        if s[-1] in street_type_lookup:
            street_suffix = s[-1]
        # add the entries into the return field
        out.append((house_number, street, street_suffix))
    return out

In [57]:
#  E WRIGHT ST / N WEIL ST,MKE
get_street_info(df['location'][5])

[(None, 'E WRIGHT', 'ST'), (None, 'N WEIL', 'ST')]

In [58]:
#  1421 N 27TH ST,MKE
get_street_info(df['location'][1])

[(1421, 'N 27TH', 'ST')]

In [59]:
df['location'].head(6)

0        7420 W GOOD HOPE RD,MKE
1             1421 N 27TH ST,MKE
2             4054 N 71ST ST,MKE
3           245 W LINCOLN AV,MKE
4            1721 W CANAL ST,MKE
5    E WRIGHT ST / N WEIL ST,MKE
Name: location, dtype: object

Some of the locations are corners of two streets and the others are single house or location addresses. Due to this it will be necessary to create a categorical feature denoting if the entry is a corner or not. The attributes of both locations will then be recorded as well. Because corner locations do not have house numbers attached these will not be included for the second location value. Overall the following features will be added:
- isCorner
- houseNumber
- primaryStreetName
- primaryStreetSuffix
- secondaryStreetName
- secondaryStreetSuffix

In [89]:
def get_street_data_as_array(location: []) -> [int, int, str, str, str, str]:
    """
    Method to take in a location that contains up to two streets and combine it into one list for output.
    The location is an array of tuples up to two in length.
    Each tuple will have 3 entries of the form (houseNumber, streetName, streetSuffix).
    If two entries are present then the location is a corner.
    The output list will be of the form [isCorner, houseNumber, primaryStreetName, primaryStreetSuffix, secondaryStreetName, secondaryStreetSuffix]
    :param location: an array of tuples up to two in length.
    :return: list will be of the form [isCorner, houseNumber, primaryStreetName, primaryStreetSuffix, secondaryStreetName, secondaryStreetSuffix]
    :auth: Grant Fass
    :since: 8 February 2022
    """
    
    if len(location) == 1:
        # is not corner
        location_vals = list(location[0])
        return [False, location_vals[0], location_vals[1], location_vals[2], None, None]
    else:
        # is corner
        primary_location_vals = list(location[0])
        secondary_location_vals = list(location[1])
        return [True, None, primary_location_vals[1], primary_location_vals[2], 
        secondary_location_vals[1], secondary_location_vals[2]]

In [90]:
get_street_data_as_array(get_street_info(df['location'][1]))

[False, 1421, 'N 27TH', 'ST', None, None]

In [91]:
get_street_data_as_array(get_street_info(df['location'][5]))

[True, None, 'E WRIGHT', 'ST', 'N WEIL', 'ST']

In [92]:
# tolist() is needed since the output is a ndarray of lists.
street_data = df["location"].map(get_street_info).map(get_street_data_as_array).tolist()

In [108]:
header = ["isCorner", "houseNumber", "primaryStreetName", "primaryStreetSuffix", "secondaryStreetName", "secondaryStreetSuffix"]
data = pd.DataFrame(street_data, columns=header)
data.head(20)

isCorner  houseNumber primaryStreetName primaryStreetSuffix  \
0      False       7420.0       W GOOD HOPE                  RD   
1      False       1421.0            N 27TH                  ST   
2      False       4054.0            N 71ST                  ST   
3      False        245.0         W LINCOLN                  AV   
4      False       1721.0           W CANAL                  ST   
5       True          NaN          E WRIGHT                  ST   
6      False       9010.0            N 95TH                  ST   
7      False        983.0          W ARTHUR                  AV   
8      False       4115.0            N 56TH                  ST   
9      False       7806.0         W HAMPTON                  AV   
10     False       9200.0      W BLUE MOUND                  RD   
11     False       3230.0          N BUFFUM                  ST   
12     False       5250.0          W LISBON                  AV   
13     False       2920.0  N VEL R PHILLIPS                  AV   
14     False       1912.0   W CLAYTON CREST                  AV   
15      True          NaN        N HUMBOLDT                  AV   
16      True          NaN            N 20TH                  ST   
17     False       5629.0           W KEEFE                  AV   
18     False        231.0         E BUFFALO                  ST   
19     False       2377.0            S 19TH                  ST   

   secondaryStreetName secondaryStreetSuffix  
0                 None                  None  
1                 None                  None  
2                 None                  None  
3                 None                  None  
4                 None                  None  
5               N WEIL                    ST  
6                 None                  None  
7                 None                  None  
8                 None                  None  
9                 None                  None  
10                None                  None  
11                None                  None  
12                None                  None  
13                None                  None  
14                None                  None  
15             E NORTH                    AV  
16           W CAPITOL                    DR  
17                None                  None  
18                None                  None  
19                None                  None

In [103]:
# data['houseNumber'] = data["houseNumber"].astype('int64')
data['primaryStreetSuffix'] = data['primaryStreetSuffix'].astype('category')
data['primaryStreetName'] = data['primaryStreetName'].astype('category')
data['secondaryStreetSuffix'] = data['secondaryStreetSuffix'].astype('category')
data['secondaryStreetName'] = data["secondaryStreetName"].astype('category')

In [104]:
data.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4027695 entries, 0 to 4027694
Data columns (total 6 columns):
 #   Column                 Non-Null Count    Dtype   
---  ------                 --------------    -----   
 0   isCorner               4027695 non-null  bool    
 1   houseNumber            3426083 non-null  float64 
 2   primaryStreetName      4027695 non-null  category
 3   primaryStreetSuffix    3844881 non-null  category
 4   secondaryStreetName    549725 non-null   category
 5   secondaryStreetSuffix  531853 non-null   category
dtypes: bool(1), category(4), float64(1)
memory usage: 58.3 MB


In [106]:
data['isCorner'].value_counts()

False    3477970
True      549725
Name: isCorner, dtype: int64

The locations have now been properly extracted into seperate features and stored in their own array with proper data types. They must now be added back in to the overall data array. It is also possible to see that there are 549,725 corner locations and 3,477,970 locations that are not corners.

In [112]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4027695 entries, 0 to 4027694
Data columns (total 9 columns):
 #   Column       Dtype         
---  ------       -----         
 0   time         datetime64[ns]
 1   location     object        
 2   district     category      
 3   nature       category      
 4   status       category      
 5   inserted_at  datetime64[ns]
 6   updated_at   datetime64[ns]
 7   point        object        
 8   call_id      int64         
dtypes: category(3), datetime64[ns](3), int64(1), object(2)
memory usage: 199.8+ MB


In [113]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4027695 entries, 0 to 4027694
Data columns (total 6 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   isCorner               bool   
 1   houseNumber            float64
 2   primaryStreetName      object 
 3   primaryStreetSuffix    object 
 4   secondaryStreetName    object 
 5   secondaryStreetSuffix  object 
dtypes: bool(1), float64(1), object(4)
memory usage: 157.5+ MB


In [117]:
df = pd.concat([df, data], join='outer', axis=1)

In [118]:
df.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4027695 entries, 0 to 4027694
Data columns (total 15 columns):
 #   Column                 Non-Null Count    Dtype         
---  ------                 --------------    -----         
 0   time                   4027695 non-null  datetime64[ns]
 1   location               4027695 non-null  object        
 2   district               3937463 non-null  category      
 3   nature                 4027695 non-null  category      
 4   status                 4027695 non-null  category      
 5   inserted_at            4027695 non-null  datetime64[ns]
 6   updated_at             4027695 non-null  datetime64[ns]
 7   point                  3958766 non-null  object        
 8   call_id                4027695 non-null  int64         
 9   isCorner               4027695 non-null  bool          
 10  houseNumber            3426083 non-null  float64       
 11  primaryStreetName      4027695 non-null  object        
 12  primaryStreetSuffix    38448

In [119]:
df.head()

time                 location district          nature  \
0 2019-05-21 15:19:03  7420 W GOOD HOPE RD,MKE        4          ACC PI   
1 2019-05-21 15:24:30       1421 N 27TH ST,MKE        3    TRAFFIC STOP   
2 2019-05-21 15:25:46       4054 N 71ST ST,MKE        7     SUBJ WANTED   
3 2019-05-21 20:46:28     245 W LINCOLN AV,MKE        2  SPECIAL ASSIGN   
4 2019-05-21 20:50:03      1721 W CANAL ST,MKE        3     TRBL W/SUBJ   

                         status         inserted_at          updated_at  \
0           Service in Progress 2019-05-21 20:51:09 2019-05-21 20:51:09   
1       City Citation(s) Issued 2019-05-21 20:57:11 2019-05-21 20:57:11   
2          Assignment Completed 2019-05-21 21:00:12 2019-05-21 21:00:12   
3           Service in Progress 2019-05-22 02:22:32 2019-05-22 02:22:32   
4  Unable to Locate Complainant 2019-05-22 02:25:33 2019-05-22 02:25:33   

                                               point    call_id  isCorner  \
0  0101000020E6100000FC7C94111793454061D971683600...  191411633     False   
1  0101000020E6100000D2AB014A4386454067C416CCA9FC...  191411672     False   
2  0101000020E610000053FFC5D8AE8B45402CAE3B270700...  191411674     False   
3  0101000020E610000078ABF8D04F804540633ABE0779FA...  191412545     False   
4  0101000020E6100000E8323509DE834540C3D7D7BAD4FB...  191412465     False   

   houseNumber primaryStreetName primaryStreetSuffix secondaryStreetName  \
0       7420.0       W GOOD HOPE                  RD                None   
1       1421.0            N 27TH                  ST                None   
2       4054.0            N 71ST                  ST                None   
3        245.0         W LINCOLN                  AV                None   
4       1721.0           W CANAL                  ST                None   

  secondaryStreetSuffix  
0                  None  
1                  None  
2                  None  
3                  None  
4                  None

The df and data DataFrame objects were combined on their columns with an inner join using [`pd.concat`](https://pandas.pydata.org/docs/reference/api/pandas.concat.html). This took df of size (4027695, 9), and data of size (4027695, 6) and updated df to be size (4027695, 15).

https://mpd.digitalpublicworks.com/?start=2019-01-05T00:00:00-06:00&end=2019-01-05T23:59:59.999999-06:00